File: connectors/dotnet/ApiGlimpse.AspNetCore/Envelope.cs
using System.Text.Json.Serialization;

namespace ApiGlimpse.AspNetCore;

/// <summary>One traffic observation in envelope v1.</summary>
public sealed class Sample
{
    [JsonPropertyName("method")]
    public string Method { get; set; } = "GET";

    [JsonPropertyName("path")]
    public string Path { get; set; } = "/";

    [JsonPropertyName("statusCode")]
    public int StatusCode { get; set; }

    [JsonPropertyName("latencyMs")]
    public long LatencyMs { get; set; }

    [JsonPropertyName("authObserved")]
    public string AuthObserved { get; set; } = "none";

    [JsonPropertyName("timestamp")]
    public string Timestamp { get; set; } = string.Empty;

    [JsonPropertyName("responseBodyCaptured")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public bool? ResponseBodyCaptured { get; set; }

    [JsonPropertyName("caller")]
    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
    public object? Caller { get; set; }

    [JsonPropertyName("request")]
    public IoSide Request { get; set; } = new();

    [JsonPropertyName("response")]
    public IoSide Response { get; set; } = new();
}

/// <summary>Request or response half of a sample.</summary>
public sealed class IoSide
{
    [JsonPropertyName("contentType")]
    public string? ContentType { get; set; }

    [JsonPropertyName("headerNames")]
    public List<string> HeaderNames { get; set; } = new();

    [JsonPropertyName("headers")]
    public Dictionary<string, string> Headers { get; set; } = new();

    [JsonPropertyName("bodyShape")]
    public object? BodyShape { get; set; }
}

/// <summary>POST /v1/samples body (version 1).</summary>
public sealed class Envelope
{
    [JsonPropertyName("version")]
    public int Version { get; set; } = Redaction.EnvelopeVersion;

    [JsonPropertyName("apiKey")]
    public string ApiKey { get; set; } = string.Empty;

    [JsonPropertyName("samples")]
    public List<Sample> Samples { get; set; } = new();

    [JsonPropertyName("sentAt")]
    public string SentAt { get; set; } = string.Empty;
}

/// <summary>Capture-time input for <see cref="EnvelopeFactory.CreateSample"/>.</summary>
public sealed class SampleInput
{
    public string Method { get; set; } = "GET";
    public string Path { get; set; } = "/";
    public int StatusCode { get; set; }
    public long LatencyMs { get; set; }
    public IDictionary<string, string>? RequestHeaders { get; set; }
    public IDictionary<string, string>? ResponseHeaders { get; set; }
    public IReadOnlyList<string>? RequestHeaderNames { get; set; }
    public IReadOnlyList<string>? ResponseHeaderNames { get; set; }
    public object? RequestBody { get; set; }
    public object? ResponseBody { get; set; }
    public bool HasRequestBody { get; set; }
    public bool HasResponseBody { get; set; }
    public bool? ResponseBodyCaptured { get; set; }
    public object? Caller { get; set; }
    public string? AuthObserved { get; set; }
    public string? Timestamp { get; set; }
}

/// <summary>Builders for envelope v1 samples.</summary>
public static class EnvelopeFactory
{
    public static Sample CreateSample(SampleInput input)
    {
        var ts = string.IsNullOrEmpty(input.Timestamp)
            ? DateTime.UtcNow.ToString("yyyy-MM-dd'T'HH:mm:ss.fff'Z'")
            : input.Timestamp!;
        var auth = string.IsNullOrEmpty(input.AuthObserved) ? "none" : input.AuthObserved!;
        var method = string.IsNullOrEmpty(input.Method) ? "GET" : input.Method.ToUpperInvariant();
        var path = string.IsNullOrEmpty(input.Path) ? "/" : input.Path;

        var reqHeaders = input.RequestHeaders ?? new Dictionary<string, string>();
        var resHeaders = input.ResponseHeaders ?? new Dictionary<string, string>();

        var reqNames = input.RequestHeaderNames is { Count: > 0 }
            ? input.RequestHeaderNames.Select(n => n.ToLowerInvariant()).ToList()
            : reqHeaders.Keys.Select(k => k.ToLowerInvariant()).ToList();
        var resNames = input.ResponseHeaderNames is { Count: > 0 }
            ? input.ResponseHeaderNames.Select(n => n.ToLowerInvariant()).ToList()
            : resHeaders.Keys.Select(k => k.ToLowerInvariant()).ToList();

        object? reqShape = input.HasRequestBody ? Redaction.ShapeBody(input.RequestBody) : null;
        object? resShape = input.HasResponseBody ? Redaction.ShapeBody(input.ResponseBody) : null;

        return new Sample
        {
            Method = method,
            Path = path,
            StatusCode = input.StatusCode,
            LatencyMs = input.LatencyMs,
            AuthObserved = auth,
            Timestamp = ts,
            ResponseBodyCaptured = input.ResponseBodyCaptured,
            Caller = input.Caller,
            Request = new IoSide
            {
                ContentType = ContentType(reqHeaders),
                HeaderNames = reqNames,
                Headers = Redaction.RedactHeaders(reqHeaders),
                BodyShape = reqShape,
            },
            Response = new IoSide
            {
                ContentType = ContentType(resHeaders),
                HeaderNames = resNames,
                Headers = Redaction.RedactHeaders(resHeaders),
                BodyShape = resShape,
            },
        };
    }

    public static Envelope CreateEnvelope(string apiKey, IEnumerable<Sample>? samples)
    {
        return new Envelope
        {
            Version = Redaction.EnvelopeVersion,
            ApiKey = apiKey,
            Samples = samples?.ToList() ?? new List<Sample>(),
            SentAt = DateTime.UtcNow.ToString("yyyy-MM-dd'T'HH:mm:ss.fff'Z'"),
        };
    }

    public static (bool Ok, string? Error) ValidateEnvelope(Envelope? env)
    {
        if (env is null) return (false, "Body must be an object");
        if (env.Version != Redaction.EnvelopeVersion) return (false, "Unsupported envelope version");
        if (env.Samples is null) return (false, "samples must be an array");
        return (true, null);
    }

    /// <summary>Classify auth from request headers (pre-redaction).</summary>
    public static string ObserveAuth(IDictionary<string, string>? headers)
    {
        if (headers is null) return "none";
        foreach (var (k, v) in headers)
        {
            if (k.Equals("authorization", StringComparison.OrdinalIgnoreCase)
                && System.Text.RegularExpressions.Regex.IsMatch(v, @"^Bearer\s+", System.Text.RegularExpressions.RegexOptions.IgnoreCase))
            {
                return "bearer";
            }
        }
        foreach (var k in headers.Keys)
        {
            if (k.Equals("cookie", StringComparison.OrdinalIgnoreCase)) return "cookie";
        }
        return "none";
    }

    /// <summary>Build SF3 caller hints. Explicit service name / X-Service-Name preferred.</summary>
    public static Dictionary<string, object?> ResolveCaller(IDictionary<string, string>? headers, string? serviceName)
    {
        var lower = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        if (headers is not null)
        {
            foreach (var (k, v) in headers) lower[k.ToLowerInvariant()] = v;
        }

        lower.TryGetValue("x-service-name", out var xSvc);
        lower.TryGetValue("x-client-name", out var xClient);
        var explicitName = FirstNonEmpty(xSvc, xClient, serviceName);
        lower.TryGetValue("user-agent", out var uaRaw);
        var ua = (uaRaw ?? string.Empty).ToLowerInvariant();

        var family = "unknown";
        if (ua.Contains("curl/", StringComparison.Ordinal) || ua == "curl")
            family = "curl";
        else if (ua.Contains("mozilla/", StringComparison.Ordinal)
                 || ua.Contains("chrome/", StringComparison.Ordinal)
                 || ua.Contains("safari/", StringComparison.Ordinal)
                 || ua.Contains("firefox/", StringComparison.Ordinal)
                 || ua.Contains("edg/", StringComparison.Ordinal))
            family = "browser";
        else if (ua.Contains("axios", StringComparison.Ordinal)
                 || ua.Contains("node-fetch", StringComparison.Ordinal)
                 || ua.Contains("go-http", StringComparison.Ordinal)
                 || ua.Contains("python-requests", StringComparison.Ordinal)
                 || ua.Contains("okhttp", StringComparison.Ordinal)
                 || ua.Contains("java/", StringComparison.Ordinal)
                 || ua.Contains("apiglimpse", StringComparison.Ordinal))
            family = "sdk";

        if (!string.IsNullOrEmpty(explicitName))
        {
            return new Dictionary<string, object?>
            {
                ["key"] = "svc:" + explicitName.ToLowerInvariant(),
                ["label"] = explicitName,
                ["serviceName"] = explicitName,
                ["userAgentFamily"] = family,
            };
        }

        return new Dictionary<string, object?>
        {
            ["key"] = "ua:" + family,
            ["label"] = "ua:" + family,
            ["serviceName"] = null,
            ["userAgentFamily"] = family,
        };
    }

    private static string? ContentType(IDictionary<string, string> headers)
    {
        foreach (var (k, v) in headers)
        {
            if (k.Equals("content-type", StringComparison.OrdinalIgnoreCase) && !string.IsNullOrEmpty(v))
            {
                return v.Split(';')[0].Trim();
            }
        }
        return null;
    }

    private static string? FirstNonEmpty(params string?[] vals)
    {
        foreach (var v in vals)
        {
            if (!string.IsNullOrWhiteSpace(v)) return v.Trim();
        }
        return null;
    }
}

File: connectors/java/src/main/java/com/apiglimpse/Envelope.java
package com.apiglimpse;

import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import java.time.Instant;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * Sample + envelope builders — mirrors packages/shared/src/envelope.js.
 */
public final class Envelope {

  private static final Pattern RE_BEARER = Pattern.compile("^Bearer\\s+", Pattern.CASE_INSENSITIVE);

  private static final ObjectMapper MAPPER = new ObjectMapper()
      .setSerializationInclusion(JsonInclude.Include.NON_NULL)
      .configure(SerializationFeature.ORDER_MAP_ENTRIES_BY_KEYS, false);

  private Envelope() {}

  public static ObjectMapper mapper() {
    return MAPPER;
  }

  public static String isoNow() {
    Instant now = Instant.now();
    // Prefer millis precision like JS toISOString
    String s = now.toString();
    if (s.length() > 23 && s.endsWith("Z")) {
      // Instant.toString may have nanos; trim to millis when present
      int dot = s.indexOf('.');
      if (dot > 0) {
        String frac = s.substring(dot + 1, s.length() - 1);
        if (frac.length() > 3) {
          return s.substring(0, dot + 1) + frac.substring(0, 3) + "Z";
        }
      }
    }
    return s;
  }

  public static Map<String, Object> createSample(SampleInput in) {
    Map<String, String> reqHeaders = in.requestHeaders != null ? in.requestHeaders : Map.of();
    Map<String, String> resHeaders = in.responseHeaders != null ? in.responseHeaders : Map.of();

    List<String> reqNames = in.requestHeaderNames != null && !in.requestHeaderNames.isEmpty()
        ? lowerNames(in.requestHeaderNames)
        : headerNames(reqHeaders);
    List<String> resNames = in.responseHeaderNames != null && !in.responseHeaderNames.isEmpty()
        ? lowerNames(in.responseHeaderNames)
        : headerNames(resHeaders);

    String method = in.method == null || in.method.isBlank() ? "GET" : in.method.toUpperCase(Locale.ROOT);
    String path = in.path == null || in.path.isBlank() ? "/" : in.path;
    String auth = in.authObserved == null || in.authObserved.isBlank() ? "none" : in.authObserved;
    String ts = in.timestamp == null || in.timestamp.isBlank() ? isoNow() : in.timestamp;

    Map<String, Object> request = new LinkedHashMap<>();
    request.put("contentType", contentType(reqHeaders));
    request.put("headerNames", reqNames);
    request.put("headers", Redaction.redactHeaders(reqHeaders));
    request.put("bodyShape", in.hasRequestBody ? Redaction.shapeBody(in.requestBody) : null);

    Map<String, Object> response = new LinkedHashMap<>();
    response.put("contentType", contentType(resHeaders));
    response.put("headerNames", resNames);
    response.put("headers", Redaction.redactHeaders(resHeaders));
    response.put("bodyShape", in.hasResponseBody ? Redaction.shapeBody(in.responseBody) : null);

    Map<String, Object> sample = new LinkedHashMap<>();
    sample.put("method", method);
    sample.put("path", path);
    sample.put("statusCode", in.statusCode);
    sample.put("latencyMs", in.latencyMs);
    sample.put("authObserved", auth);
    sample.put("timestamp", ts);
    if (in.responseBodyCaptured != null) {
      sample.put("responseBodyCaptured", in.responseBodyCaptured);
    }
    if (in.caller != null) {
      sample.put("caller", in.caller);
    }
    sample.put("request", request);
    sample.put("response", response);
    return sample;
  }

  public static Map<String, Object> createEnvelope(String apiKey, List<Map<String, Object>> samples) {
    Map<String, Object> env = new LinkedHashMap<>();
    env.put("version", Redaction.ENVELOPE_VERSION);
    env.put("apiKey", apiKey);
    env.put("samples", samples != null ? samples : new ArrayList<>());
    env.put("sentAt", isoNow());
    return env;
  }

  public static Map<String, Object> createEnvelope(
      String apiKey, List<Map<String, Object>> samples, String sentAt) {
    Map<String, Object> env = createEnvelope(apiKey, samples);
    env.put("sentAt", sentAt);
    return env;
  }

  public static ValidationResult validateEnvelope(Object body) {
    if (!(body instanceof Map<?, ?> map)) {
      return ValidationResult.fail("Body must be an object");
    }
    Object version = map.get("version");
    int v;
    if (version instanceof Number n) {
      v = n.intValue();
    } else {
      return ValidationResult.fail("Unsupported envelope version: " + version);
    }
    if (v != Redaction.ENVELOPE_VERSION) {
      return ValidationResult.fail("Unsupported envelope version: " + v);
    }
    if (!(map.get("samples") instanceof List<?>)) {
      return ValidationResult.fail("samples must be an array");
    }
    return ValidationResult.success();
  }

  /** Classify auth from request headers (pre-redaction). */
  public static String observeAuth(Map<String, String> headers) {
    if (headers == null) {
      return "none";
    }
    for (Map.Entry<String, String> e : headers.entrySet()) {
      if (e.getKey() != null
          && e.getKey().equalsIgnoreCase("authorization")
          && e.getValue() != null
          && RE_BEARER.matcher(e.getValue()).find()) {
        return "bearer";
      }
    }
    for (Map.Entry<String, String> e : headers.entrySet()) {
      if (e.getKey() != null && e.getKey().equalsIgnoreCase("cookie")) {
        return "cookie";
      }
    }
    return "none";
  }

  /** Build SF3 caller hints. Explicit service name / X-Service-Name preferred. */
  public static Map<String, Object> resolveCaller(Map<String, String> headers, String serviceName) {
    Map<String, String> lower = new LinkedHashMap<>();
    if (headers != null) {
      for (Map.Entry<String, String> e : headers.entrySet()) {
        if (e.getKey() != null) {
          lower.put(e.getKey().toLowerCase(Locale.ROOT), e.getValue());
        }
      }
    }
    String explicit = firstNonEmpty(
        lower.get("x-service-name"),
        lower.get("x-client-name"),
        serviceName);
    String ua = String.valueOf(lower.getOrDefault("user-agent", "")).toLowerCase(Locale.ROOT);
    String family;
    if (ua.contains("curl/") || "curl".equals(ua)) {
      family = "curl";
    } else if (ua.contains("mozilla/")
        || ua.contains("chrome/")
        || ua.contains("safari/")
        || ua.contains("firefox/")
        || ua.contains("edg/")) {
      family = "browser";
    } else if (ua.contains("axios")
        || ua.contains("node-fetch")
        || ua.contains("go-http")
        || ua.contains("python-requests")
        || ua.contains("okhttp")
        || ua.contains("java/")
        || ua.contains("apiglimpse")) {
      family = "sdk";
    } else {
      family = "unknown";
    }

    Map<String, Object> out = new LinkedHashMap<>();
    if (explicit != null && !explicit.isBlank()) {
      out.put("key", "svc:" + explicit.toLowerCase(Locale.ROOT));
      out.put("label", explicit);
      out.put("serviceName", explicit);
      out.put("userAgentFamily", family);
    } else {
      out.put("key", "ua:" + family);
      out.put("label", "ua:" + family);
      out.put("serviceName", null);
      out.put("userAgentFamily", family);
    }
    return out;
  }

  private static String contentType(Map<String, String> headers) {
    for (Map.Entry<String, String> e : headers.entrySet()) {
      if (e.getKey() != null && e.getKey().equalsIgnoreCase("content-type") && e.getValue() != null) {
        String ct = e.getValue().split(";", 2)[0].trim();
        return ct.isEmpty() ? null : ct;
      }
    }
    return null;
  }

  private static List<String> headerNames(Map<String, String> headers) {
    List<String> names = new ArrayList<>(headers.size());
    for (String k : headers.keySet()) {
      names.add(k.toLowerCase(Locale.ROOT));
    }
    return names;
  }

  private static List<String> lowerNames(List<String> names) {
    List<String> out = new ArrayList<>(names.size());
    for (String n : names) {
      out.add(n == null ? "" : n.toLowerCase(Locale.ROOT));
    }
    return out;
  }

  private static String firstNonEmpty(String... vals) {
    if (vals == null) {
      return null;
    }
    for (String v : vals) {
      if (v != null && !v.isBlank()) {
        return v.trim();
      }
    }
    return null;
  }

  /** Mutable builder input for {@link #createSample(SampleInput)}. */
  public static final class SampleInput {
    public String method;
    public String path;
    public int statusCode;
    public long latencyMs;
    public Map<String, String> requestHeaders;
    public Map<String, String> responseHeaders;
    public List<String> requestHeaderNames;
    public List<String> responseHeaderNames;
    public Object requestBody;
    public Object responseBody;
    public boolean hasRequestBody;
    public boolean hasResponseBody;
    public Boolean responseBodyCaptured;
    public Object caller;
    public String authObserved;
    public String timestamp;
  }

  public record ValidationResult(boolean ok, String error) {
    public static ValidationResult success() {
      return new ValidationResult(true, null);
    }

    public static ValidationResult fail(String error) {
      return new ValidationResult(false, error);
    }
  }
}
